### ISSUE TITLE
Kung Fu Master 2 WIP

### ISSUE THREAD

**Issue by arcadez2003** [libretro/mame2003-plus-libretro#1278]
@grant2258 @mahoneyt944 Any of ya's fancy taking a crack at this sometime.?? there is no big hurry as the
game has not yet been made fully playable in MAME, the encrypted blitter gfx protection would need to be
cracked and simulated or even better a dump of the MCU firstly, 

But sorting a problem i've run into will certainly get us code ready, i cant get the game to boot
with the blitter code active as yet, i think it's due to me having to guess at how to setup the hookup
between it and the rom loading table REGION_USER1 more or less and totally getting it wrong 
you'll see this at the top of the blitter handling in m62 video.

But hey it might be something else entirely that's why four more eyes on the code cant be a bad thing 
i've set it to use kungfum tiles via the mem map so it'll boot at least and that way i was able to rejig the 
proms for use in this ole core fixing the sprite colours so that's one less thing to worry about.

Basically we need to drop that and have it using the blitter and it should look like this......

https://www.youtube.com/watch?v=5qbKbJ6A_E8

My current WIP driver i ported from and romset for testing purposes included

[stuff.zip](https://github.com/libretro/mame2003-plus-libretro/files/7918793/stuff.zip)

I did ask the guy who released the dump and who had tried to emulate this 20 years ago if he
still had the old driver which he could share with me but he totally blanked me, which is a pity 
as it would have likely meant we'd not have run into this issue but what can ya do.

**Comment by mahoneyt944** [comment]
Interesting game, I'm working on some other stuff right now but I'll poke around with this game once I get this project finished up.

**Comment by grant2258** [comment]
Is this game in mame at all or are you disassembling it to see what's going on where did that video come from?

**Comment by arcadez2003** [comment]
It was dumped 20 years ago but the guy held onto it till the other day, he talked about trying to emulate it
back in the day but could not suss how to handle all the encrypted blitter gfx as things stand only the text
layer is simulated however as i said i cant get it to boot using the blitter hookup.

It's now in MAME
https://github.com/mamedev/mame/commit/ff210468d2285d6394a48b755fac0203ca6d4c93

You've bumped into this before we likely need to set up the decryption of the blitter as per MAME78 
of course im only assuming it's that i've left notes in the files about other possibles

**Comment by arcadez2003** [comment]
As i said dont knock yourselves out here as honestly he would have been better to just have submitted what
he had 20 years back as there were 20 or so MAME Dev's at that time who could have had this sorted in a day
ya know create the simulation of the blitter decryption as we're lacking a dump of the MCU which will likely
handle this.

As things stand there are maybe three in the current team who could do this it's just a case of if they have 
the time and or the motivation, so this game might not ever be playable all we can do watch this space and
see what happens, although i know of one other pcb for this game out in the wild the roms have been dumped
but as yet not shared maybe they read out the MCU which will sort all the issues full stop we'll need to wait and
see if that is the case or not and whether they're willing to now share.

Here's what the game should look like you'll see it being played a min or so in.

https://www.youtube.com/watch?v=Efr9EQkbCSQ

**Comment by grant2258** [comment]
The game is actually booting for me just to let you know will check a little more tomorrow m8 and see if there any reason its not booting for you.

![Screenshot from 2022-01-23 00-59-36](https://user-images.githubusercontent.com/6128601/150660596-48855e64-5cce-47e5-a8c8-b7c47aced736.png)

**Comment by mahoneyt944** [comment]
Looks promising

**Comment by arcadez2003** [comment]
Aye it boots because as i said i hooked it upto the fungfum tiles so it would start and i could work on the sprites

{ 0xd000, 0xdfff, kungfum_tileram_w, &m62_tileraram },

Kung Fu Master 2 does not use those tiles check the blitter write and you'll spot the tile handling but certainly 
on my build if i remove the above call from the mem map as is correct the game wont boot for me at least

**Comment by grant2258** [comment]
@arcadez2003  the reason its not playing ball without that is mt_tileram isint set in the memorymap without it adding m62_tileram = (UINT8*) auto_malloc(64*32*2); to teh videostart will allow it to boot. Im not sure where the romdata should be pointing too though

**Comment by arcadez2003** [comment]
Ah so it was that i'd ruled it out thinking it was just resizing the screen after searching for .resize on MAME
up popped a few video files we have flstory.c and ladyfrog.c and your spot on this core expects it how you've
set it up a good spot.

        kungfum2_tileram = (UINT8*) auto_malloc(64*32*2);
	return m62_start( get_kungfum2_bg_tile_info, 32, 0, 8, 8, 64, 32 );

I assume it points to the other kungfum2_tileram calls in get_kungfum2_bg_tile_info.?? im starting to seperate the
game from M62 where possible to keep the driver as close to the one we're porting from as possible i'll upload 
some new files as i've added some recent commits from MAME

**Comment by arcadez2003** [comment]
Cleaned up the code sorted a few silly mistakes my end and ported across some new commits from MAME
still not clear why ours looks like a candycane you'd get at the fair vs how it looks in current mame but i feel
we'll get there in the end :)

[more stuff.zip](https://github.com/libretro/mame2003-plus-libretro/files/7921692/more.stuff.zip)

**Comment by arcadez2003** [comment]
![kungfum2](https://user-images.githubusercontent.com/51386955/150705215-90d6906e-85ba-4568-afa4-62b4181002cc.png)

**Comment by grant2258** [comment]
Nice find works well accept its white mate if mame improves we will get more clear picture of whats going on. Think we ready to add changes anyway the controls are a lot better in this one though. I do like the original as well though

**Comment by arcadez2003** [comment]
Aye i'll submit what we have currently it's still not on a par with MAME as the text layer and the bg is borked but yeah
better to wait for it to be fixed and working in current MAME before bothering about this any further, as always
thanks for your help bigman you saved the day once again :)

**Comment by arcadez2003** [comment]
Ok code is up i've fixed the text layer but the colours are still off,

**Comment by grant2258** [comment]
just checked mame they are using the kidniki hookup will compile at some point and see how its looking there atm

**Comment by arcadez2003** [comment]
Yip bg banking i've added it before the commit

**Comment by grant2258** [comment]
think we ok for now think we might need to look at the color eprom at some point though

**Comment by grant2258** [comment]
Should have logs why it failed only possible error I could see is shadow pointer declaration in this code

**Comment by mahoneyt944** [comment]
Yes you click on the badge on the read me page and navigate to the builds and click on psp
https://git.libretro.com/libretro/mame2003-plus-libretro/-/jobs/1426201

**Comment by arcadez2003** [comment]
Looking at the error logs i see nothing in M62 but unless im mistaken TMS32010 TMS32025 CPU's are to blame.??

**Comment by grant2258** [comment]
yeap not related to this code

**Comment by mahoneyt944** [comment]
Ok thanks

**Comment by grant2258** [comment]
nothings changed in that source file either unless a new lt() function has appeared somewhere

**Comment by mahoneyt944** [comment]
Mame2003 is failing too. These are static builds so probably frontend related.

**Comment by grant2258** [comment]
src/cpu/tms32010/tms32010.c:472:static void lt(void)
src/cpu/tms32025/tms32025.c:887:static void lt(void)

change the name of one of them since they are static it shouldn't matter

**Comment by mahoneyt944** [comment]
Need to see what actually changed to cause this conflict.

**Comment by mahoneyt944** [comment]
Psp issue was caused by a tool chain update. It's been resolved for now.

**Comment by arcadez2003** [comment]
So i've been putting it off for a few days but finally got around to checking the Proms with a view to fixing 
the colours, without rabbiting on here there was a huge change to how these are handled in MAME circa 188
let's take a peek at the proms for both Kung Fu Master and Beyond Kung Fu..........

Ignoring the sprite height and timing proms as those can only be loaded the same way and in the same sequence
let's look at only the sprite and char proms for both games.........

Kung Fu Master
	ROM_REGION( 0x300, "spr_color_proms", 0 )
	ROM_LOAD( "b-1m-.bin",    0x0000, 0x0100, CRC(76c05a9c) SHA1(1f46f436a17f8c883bdd6d9804b828a81a76f880) )    /* sprite palette red component */
	ROM_LOAD( "b-1n-.bin",    0x0100, 0x0100, CRC(23f06b99) SHA1(6b3d6349f019aeab33838ae392bc3f3f89906326) )    /* sprite palette green component */
	ROM_LOAD( "b-1l-.bin",    0x0200, 0x0100, CRC(35e45021) SHA1(511b94507f41b377f38184ed9a85f34949b28d26) )    /* sprite palette blue component */

	ROM_REGION( 0x300, "chr_color_proms", 0 )
	ROM_LOAD( "g-1j-.bin",    0x0000, 0x0100, CRC(668e6bca) SHA1(cd5262b1310821ba7b12873e4db35f081d6b9df4) )    /* character palette red component */
	ROM_LOAD( "g-1f-.bin",    0x0100, 0x0100, CRC(964b6495) SHA1(76f30a65a0ded14babad2006221aa40621fb7ea1) )    /* character palette green component */
	ROM_LOAD( "g-1h-.bin",    0x0200, 0x0100, CRC(550563e1) SHA1(11edb45acba8b28a462c49956ebb1ba0a8b2ff26) )    /* character palette blue component */

Beyond Kung Fu

	ROM_REGION( 0x300, "spr_color_proms", 0 )
	ROM_LOAD( "km-b.1m", 0x0000, 0x0100, CRC(73638418) SHA1(ad3f9cf08d334e76294bd796b7f8849014f05b8e) )
	ROM_LOAD( "km-b.1n", 0x0100, 0x0100, CRC(7967dfdf) SHA1(95f5aac75ce902287c0d0ada6ee9e1e1bd9d87c0) )
	ROM_LOAD( "km-b.1l", 0x0200, 0x0100, CRC(4f44ef5c) SHA1(094e6ab24a5663208fc9d54203ed200c9e4a9fc3) )

	ROM_REGION( 0x300, "chr_color_proms", 0 )
	ROM_LOAD( "km-z-1j", 0x0000, 0x0100, CRC(1da0ad7f) SHA1(3b3ba444efa8f5481c64b3c9ef1c0ab6561c9f16) )
	ROM_LOAD( "km-z-1h", 0x0100, 0x0100, CRC(23a217c0) SHA1(723738cd8875c5bb449be2da75ef0b04cca8dd55) )
	ROM_LOAD( "km-z-1f", 0x0200, 0x0100, CRC(ec9df4f2) SHA1(702f7d536a5f79987342521c66d703153a888010) )

Both are the same barring 1.h and 1.f are in a different boot order in Beyond Kung Fu 

Let's look at how these roms are loaded in our ole core..........

Kung Fu Master

	ROM_LOAD( "g-1j-.bin",    0x0000, 0x0100, CRC(668e6bca) SHA1(cd5262b1310821ba7b12873e4db35f081d6b9df4) )	/* character palette red component */
	ROM_LOAD( "b-1m-.bin",    0x0100, 0x0100, CRC(76c05a9c) SHA1(1f46f436a17f8c883bdd6d9804b828a81a76f880) )	/* sprite palette red component */
	ROM_LOAD( "g-1f-.bin",    0x0200, 0x0100, CRC(964b6495) SHA1(76f30a65a0ded14babad2006221aa40621fb7ea1) )	/* character palette green component */
	ROM_LOAD( "b-1n-.bin",    0x0300, 0x0100, CRC(23f06b99) SHA1(6b3d6349f019aeab33838ae392bc3f3f89906326) )	/* sprite palette green component */
	ROM_LOAD( "g-1h-.bin",    0x0400, 0x0100, CRC(550563e1) SHA1(11edb45acba8b28a462c49956ebb1ba0a8b2ff26) )	/* character palette blue component */
	ROM_LOAD( "b-1l-.bin",    0x0500, 0x0100, CRC(35e45021) SHA1(511b94507f41b377f38184ed9a85f34949b28d26) )	/* sprite palette blue component */

Basically they're reversed and load in a full single sequence i assume that would mean considering the rom diffs in
Beyond Kung Fu the boot order for our driver should be as such.....

    ROM_LOAD( "km-z-1j", 0x0000, 0x0100, CRC(1da0ad7f) SHA1(3b3ba444efa8f5481c64b3c9ef1c0ab6561c9f16) )
    ROM_LOAD( "km-b.1m", 0x0100, 0x0100, CRC(73638418) SHA1(ad3f9cf08d334e76294bd796b7f8849014f05b8e) )
   	ROM_LOAD( "km-z-1h", 0x0200, 0x0100, CRC(23a217c0) SHA1(723738cd8875c5bb449be2da75ef0b04cca8dd55) )
	ROM_LOAD( "km-b.1n", 0x0300, 0x0100, CRC(7967dfdf) SHA1(95f5aac75ce902287c0d0ada6ee9e1e1bd9d87c0) )
 	ROM_LOAD( "km-z-1f", 0x0400, 0x0100, CRC(ec9df4f2) SHA1(702f7d536a5f79987342521c66d703153a888010) )
	ROM_LOAD( "km-b.1l", 0x0500, 0x0100, CRC(4f44ef5c) SHA1(094e6ab24a5663208fc9d54203ed200c9e4a9fc3) )

Of course with no way of knowing what colour the roms are it's a big assumption im making to think that just because
this one is red in Kung Fu Master

	ROM_LOAD( "g-1j-.bin",    0x0000, 0x0100, CRC(668e6bca) SHA1(cd5262b1310821ba7b12873e4db35f081d6b9df4) )	/* character palette red component */

Then this one from Beyond Kung Fu must be also as it has "j" in the rom name, it likely will though but i cant be
sure, but hey it might be i've made a mistake in the boot order or barring that the standard M62 pallete is not
correct for this game.

    ROM_LOAD( "km-z-1j", 0x0000, 0x0100, CRC(1da0ad7f) SHA1(3b3ba444efa8f5481c64b3c9ef1c0ab6561c9f16) )

Do you guys know how to scan or debug a rom to determine it's colour as the driver has no info around this and
i dunno how to do it, TBH not a big deal as we're still waiting to see if the game get's to a fully playable state firstly
as barring that correct colours wont matter anyway :)

**Comment by grant2258** [comment]
the pallet is set here https://github.com/libretro/mame2003-plus-libretro/blob/36cfc3ec93d0d322e21a355b8c4b827774d13645/src/vidhrdw/m62_vidhrdw.c#L66-L98 you might also want to check the gfxdecode as well its set to kungfumaster

**Comment by arcadez2003** [comment]
Aye i know that bigman as is the case in current MAME i was hoping the issues were confined to the proms though
without having to touch the other code but we can always cross the bridge when it comes :)

**Comment by grant2258** [comment]
I tried putting eproms in order our driver wanted them assuming the order is rgb in mame current our order last night didnt make a difference though

**Comment by arcadez2003** [comment]
i've almost got it need to check the code a little more

**Comment by arcadez2003** [comment]
Right i think we're good now https://github.com/libretro/mame2003-plus-libretro/commit/7034237eede579e8c7eba32dbd49abbaf55ef0da

**Comment by grant2258** [comment]
nice work mate I thought the pallet code would have needed a re write could have sore i replied to this yesterday couldnt have hit the comment button. Hopefully some progress is made on this one think the controls are a lot better than the original kung fu master as well

**Comment by arcadez2003** [comment]
Kung Fu Master was one my ole faves so like yourself im hoping this sequel of sorts is made fully playable, but i hear
it wont be easy as an undumped MCU handles the blitter gfx encryption so without it they're gonna have to create
a simulation from scratch which will be no easy task.

**Comment by arcadez2003** [comment]
yeah there is also a school of thought that it's not an MCU at all but either way we dont have it, and it's better for
us if it's not as IREM tended to use a I8051 for those and we dont have that CPU core, i do for my MAME84 core as i
backported it but that is another story.

My graft reflects my desire to have this playable it might not be a patch on the original though we'll see but hey
without you sorting that graphical call i'd still be trying to get this to boot with the blitter active so thanks again
big yin.

**Comment by arcadez2003** [comment]
Anyway i'll close this down for now and open up again if the game becomes playable and we're trying to hook it up here.

**Comment by mahoneyt944** [comment]
Deleted off topic psp build fail comments.
